src/js/streams.ts

```diff
@@ -0,0 +1,385 @@
+ * to tty.isatty(process.stdout.fd).
+ * If in a browser, sets stdout to write to console.log and sets isatty(stdout) to false.
+ */
+export function setDefaultStdout() {
+  if (IN_NODE) {
+    const tty = require("tty");
+    const rawstdout = (x: number) => process.stdout.write(Buffer.from([x]));
+    const isatty = tty.isatty(process.stdout.fd);
+    setRawStdout(rawstdout, isatty);
+  } else {
+    setStdout((x) => console.log(x));
+  }
+}
+
+/**
+ * Sets writes to stdout to call `stdout(line)` whenever a complete line is
+ * written or stdout is flushed. In the former case, the received line will end
+ * with a newline, in the latter case it will not.
+ *
+ * isatty(stdout) is set to false (this API buffers stdout so it is impossible
+ * to make a tty with it).
+ */
+export function setStdout(stdout: (a: string) => void) {
+  isattys.stdout = false;
+  Object.assign(ttyout_ops, make_batched_put_char(stdout));
+  refreshStreams();
+}
+
+/**
+ * Sets stdout to call `stdout(character_code)` whenever a character is written
+ * to stdout.
+ * Also sets isatty(stdout) to the value of the isatty argument (default false).
+ */
+export function setRawStdout(
+  rawstdout: (a: number) => void,
+  isatty: boolean = false,
+) {
+  isattys.stdout = isatty;
+  Object.assign(ttyout_ops, make_unbatched_put_char(rawstdout));
+  refreshStreams();
+}
+
+/**
+ * If in node, sets stderr to write directly to process.stderr and sets isatty(stderr)
+ * to tty.isatty(process.stderr.fd).
+ * If in a browser, sets stderr to write to console.warn and sets isatty(stderr) to false.
+ */
+export function setDefaultStderr() {
+  if (IN_NODE) {
+    const tty = require("tty");
+    const rawstderr = (x: number) => process.stderr.write(Buffer.from([x]));
+    const isatty = tty.isatty(process.stderr.fd);
+    setRawStderr(rawstderr, isatty);
+  } else {
+    setStderr((x) => console.warn(x));
+  }
+}
+
+/**
+ * Sets writes to stderr to call `stderr(line)` whenever a complete line is
+ * written or stderr is flushed. In the former case, the received line will end
+ * with a newline, in the latter case it will not.
+ *
+ * isatty(stderr) is set to false (this API buffers stderr so it is impossible
+ * to make a tty with it).
+ */
+export function setStderr(stderr: (a: string) => void) {
+  isattys.stderr = false;
+  Object.assign(ttyerr_ops, make_batched_put_char(stderr));
+  refreshStreams();
+}
+
+/**
+ * Sets stderr to call `stderr(character_code)` whenever a character is written
+ * to stderr.
+ * Also sets isatty(stderr) to the value of the isatty argument (default false).
+ */
+export function setRawStderr(
+  rawstderr: (a: number) => void,
+  isatty: boolean = false,
+) {
+  isattys.stderr = isatty;
+  Object.assign(ttyerr_ops, make_unbatched_put_char(rawstderr));
+  refreshStreams();
+}
+
+const textencoder = new TextEncoder();
+const textdecoder = new TextDecoder();
+
+function make_get_char(infunc: InFuncType): GetCharType {
```

Comment:
Is there a reason you named these functions in snake case?

Reply:
Well Emscripten named `get_char` and `put_char` in snake case so I figured it made sense to stick with the snake case. Since we define and use functions across three different languages following many different conventions I don't place a particularly high value on consistency. But I could change them to camel case if you prefer.